Revamp guard debug logging

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107532
* #107530
* #107516
* __->__ #107505

The new guard printout looks like this:

```
[DEBUG] GUARDS:
[DEBUG]   ___check_type_id(L['name'], 7605632)                          # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   L['name'] == '_backward_pre_hooks'                            # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   ___check_obj_id(L['self'], 139746432564960)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___check_obj_id(L['__class__'], 1451499216)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___is_grad_enabled()                                          # _dynamo/output_graph.py:346 in init_ambient_guards
[DEBUG]   not ___are_deterministic_algorithms_enabled()                 # _dynamo/output_graph.py:342 in init_ambient_guards
[DEBUG]   ___is_torch_function_enabled()                                # _dynamo/output_graph.py:350 in init_ambient_guards
[DEBUG]   utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:348 in init_ambient_guards
```

Along with the guards, we also print what line of user code caused the guard to be added, or what line of Dynamo internal code added the guard (if there is no user stack trace, which is typically the case for ambient guards.)

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ezyang committed:
Revamp guard debug logging

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2027,46 +2027,49 @@ def __init__(
                 if name in f_locals:
                     self._freevars_ids[name] = id(f_locals[name])
 
+    def init_local_index_guards_hack(self):
```
Just code motion, so I can get `init_local_index_guards_hack` in the debug print

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -346,6 +332,25 @@ def __init__(
 
+    # This gets its own helper function so guards DEBUG logs are more
+    # informative
+    def init_ambient_guards(self):
```
Just code motion, so I can get `init_ambient_guards` in the name

ezyang commented on torch/utils/_traceback.py:
```diff
@@ -130,21 +130,30 @@ def report_compile_source_on_error():
-    prefix = os.path.commonprefix([fn, os.path.join(os.path.dirname(os.path.dirname(__file__)), "")])
-    return fn[len(prefix):]
+    try:
+        prefix = os.path.commonpath([fn, base])
```
This is fixing a bug where 'test/test_foo.py' got shortened to 'est/test_foo.py'

voznesenskym replied:
lol

ezyang commented on torch/_guards.py:
```diff
@@ -148,10 +148,6 @@ class Guard:
 
-    def __post_init__(self):
-        self.stack = CapturedTraceback.extract(skip=2)
-        self.user_stack = TracingContext.extract_stack()
```
@jansel told me and I agree that the construction site of guards is not as useful as when they are installed.

voznesenskym replied:
Why not both? Both have value imo.

The way I wanted to do guard creation logging is to add a `reason` field - most initial guard association/creation sites are manual/have some reason that we hold in tribal knowledge.

ezyang replied:
It's a signal-to-noise problem. Yes, I *could* report the allocation site and the installation site but you're doubling the amount of information you have to sift through. A lot of the trickiness here is making a good decision about what to give to the user by default. That being said, I'm amenable to changing up what we record, it's not a hard change, we can tweak it as we get experience.

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -1007,7 +1007,7 @@ def convert(size_or_stride):
 
         if os.environ.get("TORCHDYNAMO_PRINT_GUARDS", None) == "1":
-            print("GUARDS", guard_body)
+            print("GUARDS\n", guard_body)
```
I am going to get to this next, but this PR already is pretty big.

ezyang committed:
Update on "Revamp guard debug logging"


The new guard printout looks like this:

```
[DEBUG] GUARDS:
[DEBUG]   ___check_type_id(L['name'], 7605632)                          # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   L['name'] == '_backward_pre_hooks'                            # if name == "special_attr":  # test/dynamo/test_misc.py:1155 in __getattribute__
[DEBUG]   ___check_obj_id(L['self'], 139746432564960)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___check_obj_id(L['__class__'], 1451499216)                   # return super().__getattribute__(name)  # test/dynamo/test_misc.py:1157 in __getattribute__
[DEBUG]   ___is_grad_enabled()                                          # _dynamo/output_graph.py:346 in init_ambient_guards
[DEBUG]   not ___are_deterministic_algorithms_enabled()                 # _dynamo/output_graph.py:342 in init_ambient_guards
[DEBUG]   ___is_torch_function_enabled()                                # _dynamo/output_graph.py:350 in init_ambient_guards
[DEBUG]   utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:348 in init_ambient_guards
```

Along with the guards, we also print what line of user code caused the guard to be added, or what line of Dynamo internal code added the guard (if there is no user stack trace, which is typically the case for ambient guards.)

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]
